library/modules/Units.cpp

```diff
@@ -1942,3 +1944,161 @@ int Units::getStressCategoryRaw(int32_t stress_level)
     }
     return level;
 }
+
+int32_t *getActionTimerPointer(df::unit_action *action) {
+    switch (action->type)
+    {
+    case unit_action_type::None:
+        break;
+    case unit_action_type::Move:
+        return &action->data.move.timer;
+    case unit_action_type::Attack:
+        if (action->data.attack.timer1 != 0) {
+            // Wind-up timer is still active, work on it
+            return &action->data.attack.timer1;
+        } else {
+            // Wind-up timer is finished, work on recovery timer
+            return &action->data.attack.timer2;
+        }
+    case unit_action_type::HoldTerrain:
+        return &action->data.holdterrain.timer;
+    case unit_action_type::Climb:
+        return &action->data.climb.timer;
+    case unit_action_type::Job:
+        return &action->data.job.timer;
+        // could also patch the unit->job.current_job->completion_timer
+    case unit_action_type::Talk:
+        return &action->data.talk.timer;
+    case unit_action_type::Unsteady:
+        return &action->data.unsteady.timer;
+    case unit_action_type::Dodge:
+        return &action->data.dodge.timer;
+    case unit_action_type::Recover:
+        return &action->data.recover.timer;
+    case unit_action_type::StandUp:
+        return &action->data.standup.timer;
+    case unit_action_type::LieDown:
+        return &action->data.liedown.timer;
+    case unit_action_type::Job2:
+        return &action->data.job2.timer;
+        // could also patch the unit->job.current_job->completion_timer
+    case unit_action_type::PushObject:
+        return &action->data.pushobject.timer;
+    case unit_action_type::SuckBlood:
+        return &action->data.suckblood.timer;
+    case unit_action_type::Jump:
+    case unit_action_type::ReleaseTerrain:
+    case unit_action_type::Parry:
+    case unit_action_type::Block:
+    case unit_action_type::HoldItem:
+    case unit_action_type::ReleaseItem:
+    case unit_action_type::Unk20:
+    case unit_action_type::Unk21:
+    case unit_action_type::Unk22:
+    case unit_action_type::Unk23:
+        break;
+    }
+    return nullptr;
+}
+
+void subtractActionTimerCore(df::unit_action *action, int32_t amount)
+{
+    int32_t *timer = getActionTimerPointer(action);
+    if (timer != nullptr && *timer > 0) {
+        double value = *timer;
+        value = max(value - amount, 1.0);
```

Comment:
This code is repetitive, can we make it less so?

Reply:
`mutateActionTimerCor(df::unit_action *action, std::function<double(double)> mutator)` maybe

Reply:
I uh, this is a bit beyond me.
It looks like you are going to pass a function to a function like in Lua and that function is going to take a double and return a double?

Reply:
The syntax can be a little hairy, but once you get it, it's a lot like Lua, as you said.

See https://github.com/DFHack/dfhack/commit/94c6bc8063b8d69c9ebd9014aaade9a8b7dc788a for an example of a function that takes and uses a lambda.

See https://github.com/DFHack/dfhack/blob/develop/plugins/overlay.cpp#L131-L135 to see an example of calling a function with a lambda expression.